### PR TITLE
Add SIP-012 style vote

### PR DESCRIPTION
**Pull request by whoabuddy**
This PR coincides with work being done in citycoins/governance#4 and citycoins/governance#5.

It implements a voting contract and "tardis" contract for viewing historical data.

### CODE REVIEW

**Comment by whoabuddy** [comment]
Update: all contracts are updated and new tests are added for both the tardis and vote contracts. Next step will be creating the mainnet versions unless we want to make the start/end block something the deployer can modify to be a bit more flexible.

Regarding cost testing at this stage, so far it looks like this will work well despite having to make a few external contract calls to gather vote data!

Initial vote as a new voter:

```
+----------------------+----------+------------+------------+
|                      | Consumed | Limit      | Percent    |
+----------------------+----------+------------+------------+
| Runtime              | 549767   | 5000000000 | 0.011%     |
+----------------------+----------+------------+------------+
| Read count           | 86       | 7750       | 1.109%     |
+----------------------+----------+------------+------------+
| Read length (bytes)  | 392601   | 100000000  | 0.392%     |
+----------------------+----------+------------+------------+
| Write count          | 5        | 7750       | 0.065%     |
+----------------------+----------+------------+------------+
| Write length (bytes) | 752      | 15000000   | 0.005%     |
+----------------------+----------+------------+------------+
```

Changing the vote:

```
+----------------------+----------+------------+------------+
|                      | Consumed | Limit      | Percent
+----------------------+----------+------------+------------+
| Runtime              | 67642    | 5000000000 | 0.001%     |
+----------------------+----------+------------+------------+
| Read count           | 10       | 7750       | 0.129%     |
+----------------------+----------+------------+------------+
| Read length (bytes)  | 9440     | 100000000  | 0.009%     |
+----------------------+----------+------------+------------+
| Write count          | 2        | 7750       | 0.026%     |
+----------------------+----------+------------+------------+
| Write length (bytes) | 405      | 15000000   | 0.003%     |
+----------------------+----------+------------+------------+

File: tests/citycoin-tardis-v2.test.ts [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+          token.ftMint(mintAmount, wallet_1)
+        ]);
+        chain.mineEmptyBlock(100);
+        chain.mineBlock([
+          token.ftMint(mintAmount, wallet_1)
+        ]);
+
+        const result1 = tardis.getHistoricalBalance(1, wallet_1).result;
+        const result2 = tardis.getHistoricalBalance(101, wallet_1).result;
+        const result3 = token.getBalance(wallet_1).result;
```
Comment (LNow) [inline tests/citycoin-tardis-v2.test.ts:43]:
```suggestion
        // arrange
        const wallet_1 = accounts.get("wallet_1")!;
        const mintAmount = 100;

        chain.mineEmptyBlock(100);
        chain.mineBlock([
          token.ftMint(mintAmount, wallet_1)
        ]);
        chain.mineEmptyBlock(100);
        chain.mineBlock([
          token.ftMint(mintAmount, wallet_1)
        ]);

        // act
        const result1 = tardis.getHistoricalBalance(1, wallet_1).result;
        const result2 = tardis.getHistoricalBalance(101, wallet_1).result;
        const result3 = token.getBalance(wallet_1).result;
```
Minting coins is something that is part of test arrange activity.

File: tests/citycoin-tardis-v2.test.ts [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+          token.ftMint(mintAmount, wallet_1)
+        ]);
+        chain.mineEmptyBlock(100);
+        chain.mineBlock([
+          token.ftMint(mintAmount, wallet_1)
+        ]);
+
+        const result1 = tardis.getHistoricalSupply(1).result;
+        const result2 = tardis.getHistoricalSupply(101).result;
+        const result3 = token.getTotalSupply().result;
```
Comment (LNow) [inline tests/citycoin-tardis-v2.test.ts:70]:
```suggestion
        // arrange
        const wallet_1 = accounts.get("wallet_1")!;
        const mintAmount = 100;

        chain.mineEmptyBlock(100);
        chain.mineBlock([
          token.ftMint(mintAmount, wallet_1)
        ]);
        chain.mineEmptyBlock(100);
        chain.mineBlock([
          token.ftMint(mintAmount, wallet_1)
        ]);
        
        // act
        const result1 = tardis.getHistoricalSupply(1).result;
        const result2 = tardis.getHistoricalSupply(101).result;
        const result3 = token.getTotalSupply().result;
```

File: tests/citycoin-vote-v1.test.ts [hunk 3]
```diff
@@ -0,0 +1,874 @@
...
+      it("fails with ERR_PROPOSAL_NOT_ACTIVE when called after proposal is active", () => {
+        // arrange
+        const wallet = accounts.get("wallet_1")!;
+        const deployer = accounts.get("deployer")!;
+        const startHeight = 8500;
+        const endHeight = 10600;
+        chain.mineBlock([
+          vote.initializeContract(startHeight, endHeight, deployer)
+        ]);
+        chain.mineEmptyBlock(VoteModel.VOTE_START_BLOCK + VoteModel.VOTE_END_BLOCK + 1);
```
Comment (LNow) [inline tests/citycoin-vote-v1.test.ts:123]:
```suggestion
        chain.mineEmptyBlockUntil(VoteModel.VOTE_END_BLOCK + 1);
```
Moving chain tip to `VOTE_END_BLOCK + 1` (10601) feels a bit more appropriate than `VOTE_START_BLOCK + VOTE_END_BLOCK + 1` (19101).

**Review by LNow — CHANGES_REQUESTED** [review]

File: tests/citycoin-vote-v1.test.ts (hunk 3 above)
Comment (whoabuddy) [reply tests/citycoin-vote-v1.test.ts:123]: Yes, good catch, I think my over-complicated intent was actually `VOTE_START_BLOCK + (VOTE_END_BLOCK - VOTE_START_BLOCK) + 1`

**Review by LNow — APPROVED** [review]

**Pull request merged**